src/gluonts/model/forecast_generator.py

```diff
@@ -0,0 +1,222 @@
+        logging.info(msg)
+        LOG_CACHE.add(msg)
+
+
+def _extract_instances(x: Any) -> Any:
+    """
+    Helper function to extract individual instances from batched
+    mxnet results.
+
+    For a tensor `a`
+      _extract_instances(a) -> [a[0], a[1], ...]
+
+    For (nested) tuples of tensors `(a, (b, c))`
+      _extract_instances((a, (b, c)) -> [(a[0], (b[0], c[0])), (a[1], (b[1], c[1])), ...]
+    """
+    if isinstance(x, (np.ndarray, mx.nd.NDArray)):
+        for i in range(x.shape[0]):
+            # yield x[i: i + 1]
+            yield x[i]
+    elif isinstance(x, tuple):
+        for m in zip(*[_extract_instances(y) for y in x]):
+            yield tuple([r for r in m])
+    elif x is None:
+        while True:
+            yield None
+    else:
+        assert False
+
+
+class ForecastGenerator:
+    """
+    Classes used to bring the output of a network into a class.
+    """
+
+    def __call__(
+        self,
+        inference_data_loader: InferenceDataLoader,
+        prediction_net: BlockType,
+        input_names: List[str],
+        freq: str,
+        output_transform: Optional[OutputTransform],
+        num_eval_samples: Optional[int],
+        **kwargs
+    ) -> Iterator[Forecast]:
+        raise NotImplementedError()
+
+
+class DistributionForecastGenerator(ForecastGenerator):
+    @validated()
+    def __init__(self, distr_output: DistributionOutput) -> None:
+        self.distr_output = distr_output
+
+    def __call__(
+        self,
+        inference_data_loader: InferenceDataLoader,
+        prediction_net: BlockType,
+        input_names: List[str],
+        freq: str,
+        output_transform: Optional[OutputTransform],
+        num_eval_samples: Optional[int],
+        **kwargs
+    ) -> Iterator[DistributionForecast]:
+        for batch in inference_data_loader:
+            inputs = [batch[k] for k in input_names]
+            outputs = prediction_net(*inputs)
+            if output_transform is not None:
+                outputs = output_transform(batch, outputs)
+            if num_eval_samples:
+                log_once(
+                    "Forecast is not sample based. Ignoring parameter `num_eval_samples` from predict method."
+                )
+
+            distributions = [
+                self.distr_output.distribution(*u)
+                for u in _extract_instances(outputs)
+            ]
+
+            i = -1
+            for i, distr in enumerate(distributions):
+                yield DistributionForecast(
+                    distr,
+                    start_date=batch["forecast_start"][i],
+                    freq=freq,
+                    item_id=batch[FieldName.ITEM_ID][i]
+                    if FieldName.ITEM_ID in batch
+                    else None,
+                    info=batch["info"][i] if "info" in batch else None,
+                )
+            assert i + 1 == len(batch["forecast_start"])
+
+
+class QuantileForecastGenerator(ForecastGenerator):
```

Comment:
This class looks a lot like `DistributionForecastGenerator`, except for minor differences in the way forecast objects are constructed. We can merge it this way for the time being, I guess, but in the future it would be nice to consolidate `QuantileForecast[Generator]` into `DistributionForecast[Generator]` through an appropriate `Distribution` subtype. (The same could apply to `SampleForecast[Generator]` but I see there's a bit more logic there so that might require some thinking)

Reply:
yes, that might be nice